### ISSUE TITLE
GCS gfile operations fail in TF 2.17.0rc0 and 2.18 nightly when not running in GCP

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#69789]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.17.0rc0, 2.17.0-dev20240320, 2.17.0.dev20240528, 2.18.0.dev20240613, 2.18.0.dev20240617

### Custom code

No

### OS platform and distribution

Debian bookworm, macos 14.5

### Python version

3.11

### Current behavior?

When trying to run GCS operations with `tf.io.gfile` on nightly 2.17 or 2.18 anywhere outside of a GCP VM the command hangs and eventually fails after 10 retries with the error message as below.

I can't seem to reproduce this issue on either colab or a GCP VM. But it will consistently fail locally on my mac, inside a `python:3.11` docker container, on GitHub actions or inside a kaggle notebook. The same code works fine with TensorFlow 2.16 so I don't think this is due to my local setup.

It also seems like other people are running into this with the latest TF nightly: https://github.com/tensorflow/datasets/issues/5360

Would be great to get this fixed before the 2.17 stable release.

### Standalone code to reproduce the issue

```python
import tensorflow as tf

tf.io.gfile.exists("gs://tfds-data/dataset_info/mnist/3.0.1/dataset_info.json")
```

### Relevant log output

```shell
2024-06-14 15:45:30.081439: W external/local_tsl/tsl/platform/cloud/google_auth_provider.cc:184] All attempts to get a Google authentication bearer token failed, returning an empty token. Retrieving token from files failed with "NOT_FOUND: Could not locate the credentials file.". Retrieving token from GCE failed with "FAILED_PRECONDITION: Error executing an HTTP request: libcurl code 6 meaning 'Couldn't resolve host name', error details: Could not resolve host: metadata.google.internal".
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.11/site-packages/tensorflow/python/lib/io/file_io.py", line 290, in file_exists_v2
    _pywrap_file_io.FileExists(compat.path_to_bytes(path))
tensorflow.python.framework.errors_impl.AbortedError: All 10 retry attempts failed. The last failure: Error executing an HTTP request: HTTP response code 301 with body '<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.googleapis.com/storage/v1/b/tfds-data/o/dataset_info%2Fmnist%2F3.0.1%2Fdataset_info.json?fields=size%2Cgeneration%2Cupdated">here</A>.
</BODY></HTML>
'
	 when reading metadata of gs://tfds-data/dataset_info/mnist/3.0.1/dataset_info.json
```

**Comment by lgeiger** [comment]
The only commit that comes to mind that might influence this is a4e3b786447042f4aa6ad5649a7fef9c4b40cee7, but I can't easily test this since tf-nightly wheels on pypi older than this date are not available.

**Comment by mihaimaruseac** [comment]
@learning-to-play , this is another regression in TF 2.17 RC0

**Comment by learning-to-play** [comment]
Please consider filing an issue in TensorFlow IO. To my knowledge there are similar issues already reported to TensorFlow IO.

FYI @yongtang @monicadsong

**Comment by lgeiger** [comment]
> Please consider filing an issue in TensorFlow IO. To my knowledge there are similar issues already reported to TensorFlow IO.

@learning-to-play I already did at https://github.com/tensorflow/io/issues/2016.
But I'm not sure that it's related to TensorFlow IO since I can only reproduce this issue with TF 2.17 but not TF 2.16 even when both of them use the same `tensorflow_io_gcs_filesystem` package version.

**Comment by learning-to-play** [comment]
@aaudiber Could you PTAL and help route to the right person?

**Comment by mihaimaruseac** [comment]
This is from tsl layer, not TF-IO.

**Comment by belitskiy** [comment]
Indeed, using a wheel from Feb 20 works:
`gs://tensorflow-public-build-artifacts/prod/tensorflow/official/release/nightly/linux_x86_cpu/wheel_py311/391/20240220-042013/github/tensorflow/build_output/tf_nightly_cpu-2.17.0.dev20240220-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl`

(there's more wheels in that bucket)

I've been able to reproduce  this in one place, but it does work in a Docker container on a non-VM PC, so it's not necessarily VM-related.

**Comment by mihaimaruseac** [comment]
Probably could be "reverted" via copybara

**Comment by lgeiger** [comment]
Thanks for looking into it and reproducing the issue. Also thanks for sharing the link to the older tf-nightly wheels, that's very useful to know.

**Comment by lgeiger** [comment]
This has been fixed in the latest nightly and on 2.17.0rc1 🎉
